File: classification/naive_bayes.R
# This script creates and evaluates an Naive Bayes model in R for the 'iris' data.
# This iteration is only to get the model up and running, so there is no feature engineering and parameter tuning.

# Load packages.
library(data.table)
library(e1071)

# Import data.
train = data.table(read.csv('train.csv',
                            header = T))
test = data.table(read.csv('test.csv', 
                           header = T))

#####
# CREATE MODEL AND PREDICTIONS.
  
  # Set seed to ensure reproducibility.
  set.seed(1)
  
  # Define model.
  model = Species ~ Sepal.Length + Sepal.Width + Petal.Length + Petal.Width
  
  # Create model and generate predictions for training set.
  nb = naiveBayes(model, 
                  data = train)
  
  # Generate predictions for training and test data.
  train$pred = data.table(predict(nb, 
                                   newdata = train, 
                                   type = 'class'))
  test$pred = data.table(predict(nb, 
                                 newdata = test, 
                                 type = 'class'))

#####
# CHECK TRAINING SET ACCURACY.
  
  # View training accuracy.
  print('Training Accuracy:')
  table(train$Species == train$pred)
  prop.table(table(train$Species == train$pred))

#####
# CROSS VALIDATION.
    
  # Create 'tune.control' object.
  tune_control = tune.control(cross = 2)
  
  # Execute cross validation.
  t = tune(method = 'naiveBayes', 
           model, 
           data = train, 
           size = 2, 
           tunecontrol = tune_control)
  
  # View estimated test set error computed by cross validation.
  print('Cross Validation Error: ')
  print(t)

#####
# CHECK TEST SET ACCURACY.

  # View test set accuracy.
  print('Test Set Accuracy:')
  table(test$Species == test$pred)
  prop.table(table(test$Species == test$pred))

File: classification/naive_bayes.py
# This script creates and evaluates a Naive Bayes model in Python for the 'iris' data.
# This iteration is only to get the model up and running, so there is no feature engineering and parameter tuning.

# Import modules.
import os
import pandas as pd
from sklearn import cross_validation
from sklearn.naive_bayes import GaussianNB

# Import data.
# The data files are located in the 'Python' branch file, NOT the 'classification' or 'regression' folders.
# The commands below assume the working directory is set to the 'classification' folder.
os.chdir('..')
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# CREATE MODEL AND PREDICTIONS.

# Create classifier.
nb = GaussianNB()

# Fit classifier on training data.
nb.fit(train[['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width']],
       train['Species'])

# Generate predictions for training and test data.
train['pred'] = nb.predict(train[['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width']])
test['pred'] = nb.predict(test[['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width']])

# CHECK TRAINING SET ACCURACY.

# Compute training set accuracy.
print('Training Set Accuracy :', sum(train['Species'] == train['pred']) / len(train['pred']))

# CROSS VALIDATION.

# Get cross validation scores.
cv_scores = cross_validation.cross_val_score(nb,
                                             train[['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width']],
                                             train['Species'],
                                             cv = 2)

# Take the mean accuracy across all cross validation segments.
print('Cross Validation Accuracy: ', cv_scores.mean())
                                            
# CHECK TEST SET ACCURACY.

# Compute test set accuracy.
print('Test Set Accuracy :', sum(test['Species'] == test['pred']) / len(test['pred']))
